marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/WriteStepRunner.java

```diff
@@ -426,23 +478,102 @@ private Job runIngester(Job job) {
     }
 
-    private void addToBatcher(File file, Format fileFormat, DocumentMetadataHandle metadataHandle) throws  IOException{
+    private void addToBatcher(File file, Format fileFormat) throws  IOException{
         FileInputStream docStream = new FileInputStream(file);
-        InputStreamHandle handle = new InputStreamHandle(docStream);
-        handle.setFormat(fileFormat);
-/*        String absPath = file.getAbsolutePath();
-        String[] inputs = outputUriReplace.split(",", 2);
-        String stringToBeReplaced = inputs[0];
-        String */
-        writeBatcher.add(file.getAbsolutePath(), metadataHandle, handle);
+        if(inputFormat.equalsIgnoreCase("csv")) {
+            JacksonCSVSplitter splitter = new JacksonCSVSplitter();
+            try {
+                Stream<JacksonHandle> contentStream = splitter.split(docStream);
+                Path dirPath = Paths.get(filePath);
+                String uri = null;
+                if(! dirPath.isAbsolute()) {
+                    File csvFile = new File(hubConfig.getProjectDir(), dirPath.toString());
+                    uri = generateAndEncodeURI(csvFile.getParentFile().getAbsolutePath());
+
+                }
+                Stream<DocumentWriteOperation> documentStream =  DocumentWriteOperation.from(
+                    contentStream, DocumentWriteOperation.uriMaker(outputURIReplace(uri)+"/%s.json"));
+                writeBatcher.addAll(documentStream);
+            } catch (Exception e) {
+               throw new RuntimeException(e);
+            }
+        }
+        else {
+            InputStreamHandle handle = new InputStreamHandle(docStream);
+            handle.setFormat(fileFormat);
+            try {
+                writeBatcher.add(outputURIReplace(generateAndEncodeURI(file.getAbsolutePath())), handle);
+            } catch (URISyntaxException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private String generateAndEncodeURI(String path) throws  URISyntaxException {
+        URI uri = new URI(null, null, null, 0, path, null, null);
+        return uri.toString();
+    }
+
+    private void applyPermissions(String permissions, DocumentMetadataHandle metadataHandle) {
+        String[] perms = permissions.split(",");
+        if (perms != null && perms.length > 0) {
+            if (perms.length % 2 != 0) {
+                throw new IllegalArgumentException(
+                    "Permissions are expected to be in <role, capability> pairs.");
+            }
+            int i = 0;
+            while (i + 1 < perms.length) {
+                String roleName = perms[i++];
+                if (roleName == null || roleName.isEmpty()) {
+                    throw new IllegalArgumentException(
+                        "Illegal role name: " + roleName);
+                }
+                String perm = perms[i].trim();
+                if (!perm.equalsIgnoreCase(DocumentMetadataHandle.Capability.READ.toString()) &&
```

Comment:
we had better use if (!DocumentMetadataHandle.Capability.READ.toString().equalsIgnoreCase(perm)) ...